[Ad hoc metrics for OPS] Add an autocomplete input box for changing the tenant

yaacov opened the pull request:
**Description**
Refactor #12165
- [x] Add auto-complete text widget to change tenant in ad-hoc metrics page
- [x] Add icon to identify tenant metric type

**Screenshots**
_Navigation from container provider summary page_
![screenshot-localhost 3000-2017-01-20-13-59-50](https://cloud.githubusercontent.com/assets/2181522/22148799/ed84a210-df18-11e6-891c-c119cbdd81b9.png)


_Usage of the auto-complete feature_
![screenshot-localhost 3000-2017-01-18-16-32-38](https://cloud.githubusercontent.com/assets/2181522/22067903/d4116aa8-dd9b-11e6-876f-e83180ac6ebc.png)

(converted from ManageIQ/manageiq#13016)

yaacov commented:
This is converted from ManageIQ/manageiq#13016 :

Currently the available tenants are hard-coded `["_ops", "_system"],` this PR make it possible to use any tenant.

@jeff-phillips-18 @simon3z @himdel @moolitayer @zeari please review / merge

martinpovolny commented:
@himdel : please, take a look at the `$element` usage.

himdel commented on app/assets/javascripts/controllers/container_live_dashboard/container_live_dashboard_controller.js:
```diff
@@ -68,11 +70,10 @@ miqHttpInject(angular.module('containerLiveDashboard', ['ui.bootstrap', 'pattern
-    var doRefreshTenant = function (action) {
-      $scope.tenant = action.tenant;
+    $scope.doRefreshTenant = function() {
+      $scope.tenant = $element.find("#ad-hoc-tenant").val();
```
So.. what's the advantage of this over reading `$scope.tenant`? If feels *wrong*, since that element has `ng-model="tenant"` and is in the same controller. So the whole line looks like it shouldn't even be needed there.. so why is it?

himdel commented on spec/javascripts/controllers/container_live_dashboard/container_live_dashboard_controller_spec.js:
```diff
@@ -9,8 +9,12 @@ describe('containerLiveDashboardController', function() {
     var $window = {location: { pathname: '/ems_container/42' }};
+    var $element = function() {
+      this.find = function() { return "_ops" };
+    };
```
the mock doesn't make sense.. you're calling `$element.find("#whatever").val()` .. soo.. `"_ops".val()`? ;)

yaacov commented:
@himdel hi, made this changes:

a. using `this` instead of `$scope`
b. not using `$element` to get the tenant value already mapped to `this` using angular's `ng-model` ...

yaacov commented:
@himdel @martinpovolny @mzazrivec please re-review

simon3z commented:
@yaacov what is current behavior (without this PR)? We were just showing counters?

yaacov commented:
> what is current behavior (without this PR)? We were just showing counters?

@simon3z we are just showing gauges

p.s.

a. this is because the ops team collectors only send data to gauges (and strings, but we do not graph them).
b. we need counters because some of the metrics is collected from counters and it is nice to understand what raw data we have for the metrics.

himdel commented on app/assets/javascripts/controllers/container_live_dashboard/container_live_dashboard_controller.js:
```diff
@@ -1,88 +1,90 @@
-  .controller('containerLiveDashboardController', ['$scope', 'pfViewUtils', '$http', '$interval', '$timeout', '$window',
-  function ($scope, pfViewUtils, $http, $interval, $timeout, $window) {
-    $scope.tenant = '_ops';
+  .controller('containerLiveDashboardController', ['pfViewUtils', '$http', '$interval', '$timeout', '$window',
```
`pfViewUtils`, `$interval`, `$timeout` are all unused..

himdel commented on app/assets/javascripts/controllers/container_live_dashboard/container_live_dashboard_controller.js:
```diff
@@ -185,59 +187,66 @@ miqHttpInject(angular.module('containerLiveDashboard', ['ui.bootstrap', 'pattern
 
-    $scope.timeRanges = [
+    dash.getTenants = function(include) {
+      dash.tenantChanged = true;
```
Doesn't quite make sense to set `tenantChanged` here (while downloading a list of options), does it? What about watching `dash.tenant` for changes instead?

simon3z commented:
> @simon3z we are just showing gauges
> 
> p.s.
> 
> a. this is because the ops team collectors only send data to gauges (and strings, but we do not graph them).
> b. we need counters because some of the metrics is collected from counters and it is nice to understand what raw data we have for the metrics.

@yaacov @lucasponce did we have an API to show all the series (counters/gauges)?

Anyway UI-wise I am not sure if it makes sense for the user to filter on just counters or gauges. He may be familiar with what he's looking for (e.g. cpu-related metrics) but he probably doesn't know whether the series he's looking for is a counter or a gauge.

Can't we list them all? (Or anyway have the filter being "Both" by default?).

yaacov commented:
> Can't we list them all? 

@lucasponce @simon3z yes we can :-) we simply do two requests to hawkular.
This can look a bit strange e.g. memory/usage (the counter) and memory/usage (the gauge) one after the other ... but we can add a flag or some other indication.

lucasponce commented:
> @lucasponce @simon3z yes we can :-) we simply do two requests to hawkular.

It is also possible to list several metrics in a single query. It is something the middleware provider does after asking the metrics that are available per resource from the inventory.

Code used is listed here (in case some of this idea can be reused on this use case):
https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/hawkular/middleware_manager/live_metrics_capture.rb#L61

lucasponce commented:
> @lucasponce :+1: is their a way to list all the keys (id's) by tags, we use:

The metrics API offers a general endpoint
http://www.hawkular.org/docs/rest/rest-metrics.html#GET__metrics
Which I am not sure if the query commented is available in the current client code.

There is this one
https://github.com/hawkular/hawkular-client-ruby/blob/master/lib/hawkular/metrics/metric_api.rb#L133

Could you open an issue and some colleague can take a look at it ?

yaacov commented:
> Could you open an issue and some colleague can take a look at it ?

@lucasponce It's not that important ( to me :-) ), the ops team only need gauges (since they only push to gauges) Doing two queries in this context is not problematic and even helpful , since we will have a lot of duplicate id's, if we get them in two different queries, it will be easier to sort them out.

simon3z commented:
> @lucasponce It's not that important ( to me :-) ), the ops team only need gauges (since they only push to gauges)

@yaacov remember that the scope is not only ops. We want to cover heapster as well.

yaacov commented:
@simon3z now showing always both counter and gauges

@himdel Thanks 👍 , fixed the extra arguments and the strange on_chnage

simon3z commented:
> @simon3z now showing always both counter and gauges

@yaacov I suppose you can't mix counters and gauges in the same graph. Does it even make sense?

In case someone selects gauges and counters to graph do we want to keep them separate?

What about strings? Will patternfly crash in case there are strings? (Not for this PR, in case you need a fix here you could do it in another PR).

yaacov commented:
@simon3z:

> I suppose you can't mix counters and gauges in the same graph. Does it even make sense?

a. we can mix counters and gauges in the same graph
b. i do not think it make sense, but if someone wants to, we don't care :-)

>What about strings? 

We do not show strings in the list (only gauges and counters), so users can't try to graph them ..

yaacov commented:
@simon3z is the feature complete ( we list and chart both counters and gauges ) ?
@himdel is the javascript merge ready ( fixed the extra arguments and the strange on_chnage ) ?

simon3z commented on app/services/hawkular_proxy_service.rb:
```diff
@@ -33,7 +41,15 @@ def data(query)
     tags = nil if tags == {}
-    client.gauges.query(tags).compact.map { |m| m.json if m.json }.sort { |a, b| a["id"].downcase <=> b["id"].downcase }
+
+    if @params['type'] == 'both'
```
@yaacov `both`? This doesn't look like something you would use in an API.
When the `type` parameter is missing you would query all types.

simon3z commented on app/services/hawkular_proxy_service.rb:
```diff
@@ -8,10 +8,16 @@ def initialize(provider_id, controller)
+    @cli = ManageIQ::Providers::Kubernetes::ContainerManager::MetricsCapture::HawkularClient.new(@ems, @tenant)
   end
 
   def client
```
@yaacov is this still necessary? If it's used only by `metric_definitions` I think it can be dropped.

yaacov replied:
@simon3z the `client` method is used in `metric_definitions` and `get_data`

https://github.com/ManageIQ/manageiq-ui-classic/pull/30/files#diff-deb60de663c529247f745d1a17a8e551R49
and in
https://github.com/ManageIQ/manageiq-ui-classic/pull/30/files#diff-deb60de663c529247f745d1a17a8e551R66

without it, I will have to add two `if .. else` that will not add to the readability of this functions.

simon3z commented on app/services/hawkular_proxy_service.rb:
```diff
@@ -45,12 +61,27 @@ def get_data(id)
+
+  def tenants
+    tenants = @cli.hawkular_client.http_get('/tenants')
+    tenants = if @params['include'].blank?
```
@yaacov what do you use include for? To limit the list of tenant ids to tenants that you already know about?

yaacov replied:
The auto-complete widget is asking for tenants that has a name that include  `@params['include']`

https://github.com/ManageIQ/manageiq-ui-classic/pull/30/files#diff-deb60de663c529247f745d1a17a8e551R81

yaacov replied:
@simon3z it make more sense to query only relevant tenants from hawkular, but hawkular does not know how to do this :-(
The next best thing is to reduce the list on the rails side.
Sending all the tenants ( can be a lot ) to the javascript, and reducing the list on the client side is also an option.

yaacov commented:
@simon3z , removed the `both` from the api, now if no `type` is specified [ or the type specified is unavailable ] we give all available types, and if we specify an available `type` we give only data relevant to this type.

simon3z commented:
@yaacov can you check travis?

yaacov commented:
> can you check travis?

@simon3z travis was unrelated, fixed now

simon3z commented:
LGTM :+1:  cc @himdel @chessbyte

himdel commented on app/assets/javascripts/controllers/container_live_dashboard/container_live_dashboard_controller.js:
```diff
@@ -101,7 +99,8 @@ miqHttpInject(angular.module('containerLiveDashboard', ['ui.bootstrap', 'pattern
 
     var getLatestData = function(item) {
-      var params = '&query=get_data&metric_id=' + item.id + '&limit=5&order=DESC';
+      var params = '&query=get_data&type=' + item.type + 's&metric_id=' + item.id +
```
> `+ item.type + 's&metric_id=' +`

Guessing a typo? (that `s` before `&metric_id`)

himdel commented on app/assets/javascripts/controllers/container_live_dashboard/container_live_dashboard_controller.js:
```diff
@@ -163,13 +162,13 @@ miqHttpInject(angular.module('containerLiveDashboard', ['ui.bootstrap', 'pattern
       var starts = ends - diff;
       var bucket_duration = parseInt(diff / 1000 / 200); // bucket duration is in seconds
-      var params = '&query=get_data&metric_id=' + metric_id + '&ends=' + ends +
+      var params = '&query=get_data&type=' + metric_type + 's&metric_id=' + metric_id + '&ends=' + ends +
```
..and here.. or are you trying to pluralize the metric_type?

himdel replied:
If so .. I think doing it here is a Bad Idea, but I don't have any specific complaint about it, so feel free to ignore

himdel commented:
@yaacov please add where to find this in the UI to the PR description

yaacov commented:
@himdel 👍 

> are you trying to pluralize the metric_type?

i do not remember why I added the pluralizer, so just removed it :-)

> please add where to find this in the UI to the PR description

added a screen-shot of  where to find this in the UI to the PR description

himdel commented:
@yaacov Thanks for the quick response! :+1:

Code looks good, testing in the ui..

himdel commented:
Tested in the UI, works just fine, merging :).

My only remaining concern is from a UX perspective, you have to confirm twice, first time by clicking a button to the **left** of the input, the second by clicking a button to the **right** of a combobox. But that can be addressed separately if needed..

yaacov commented:
> My only remaining concern is from a UX perspective, you have to confirm twice, first time by clicking a button to the left of the input, the second by clicking a button to the right of a combobox. But that can be addressed separately if needed..

Adding @serenamarie125 to this

cc @himdel @simon3z
